src/coreclr/src/vm/ipcstreamfactory.cpp

```diff
@@ -0,0 +1,207 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "diagnosticsprotocol.h"
+#include "ipcstreamfactory.h"
+
+#ifdef FEATURE_PERFTRACING
+
+CQuickArrayList<IpcStreamFactory::ConnectionState*> IpcStreamFactory::s_rgpConnectionStates = CQuickArrayList<IpcStreamFactory::ConnectionState*>();
+Volatile<bool> IpcStreamFactory::s_isShutdown = false;
+
+bool IpcStreamFactory::ClientConnectionState::GetIpcPollHandle(IpcStream::DiagnosticsIpc::IpcPollHandle *pIpcPollHandle, ErrorCallback callback)
+{
+    if (_pStream == nullptr)
```

Comment:
Nit: The code can be a little more compact and easy to follow if the error path immediately follows the API call

````
    if(_pStream == nullptr)
    {
        // cache is empty, reconnect, e.g., there was a disconnect
        IpcStream *pConnection = _pIpc->Connect(callback);
        if (pConnection == nullptr)
        {
            if (callback != nullptr)
                callback("Failed to connect to client connection", -1);
            return false;
        }
        if (!DiagnosticsIpc::SendIpcAdvertise_V1(pConnection))
        {
            if (callback != nullptr)
                callback("Failed to send advertise message", -1);
            delete pConnection;
            return false;
        }
        _pStream = pConnection;
   }
   *pIpcPollHandle = { nullptr, _pStream, 0, this };
    return true;
````